plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs

```diff
@@ -929,6 +931,44 @@ compileExpr e = traceCompilation 2 ("Compiling expr:" GHC.<+> GHC.ppr e) $ do
   caseIntegerName <- lookupGhcName 'Builtins.caseInteger
   listIndexId <- lookupGhcId '(PlutusTx.List.!!)
 
+  let
+    compileMkNil
+      :: (CompilingDefault uni fun m ann)
+      => GHC.Type
+      -> m (PIRTerm uni fun)
+    compileMkNil ty =
+      let
+        tyToUniverse :: GHC.Type -> Maybe (SomeTypeIn' PLC.DefaultUni)
+        tyToUniverse (GHC.TyConApp tyCon [tyArg1, tyArg2])
+          | tyCon == builtinPairTyCon = do
+              (SomeTypeIn' a) <- tyToUniverse tyArg1
+              (SomeTypeIn' b) <- tyToUniverse tyArg2
+              pure $ SomeTypeIn' (PLC.DefaultUniPair a b)
+          | otherwise = Nothing
+        tyToUniverse (GHC.TyConApp tyCon [tyArg1])
+          | tyCon == builtinListTyCon = do
+              (SomeTypeIn' a) <- tyToUniverse tyArg1
+              pure $ SomeTypeIn' (PLC.DefaultUniList a)
+          | tyCon == builtinArrayTyCon = do
+              (SomeTypeIn' a) <- tyToUniverse tyArg1
+              pure $ SomeTypeIn' (PLC.DefaultUniArray a)
+          | otherwise = Nothing
+        tyToUniverse (GHC.TyConApp tyCon [])
+          | tyCon == GHC.integerTyCon || tyCon == builtinIntegerTyCon =
+            pure $ SomeTypeIn' PLC.DefaultUniInteger
+          | tyCon == builtinByteStringTyCon = pure $ SomeTypeIn' PLC.DefaultUniByteString
+          | tyCon == GHC.boolTyCon = pure $ SomeTypeIn' PLC.DefaultUniBool
+          | tyCon == builtinDataTyCon = pure $ SomeTypeIn' PLC.DefaultUniData
+          | tyCon == builtinBLS12_G1_TyCon = pure $ SomeTypeIn' PLC.DefaultUniBLS12_381_G1_Element
+          | tyCon == builtinBLS12_G2_TyCon = pure $ SomeTypeIn' PLC.DefaultUniBLS12_381_G2_Element
+          | tyCon == builtinValueTyCon = pure $ SomeTypeIn' PLC.DefaultUniValue
+          | otherwise = Nothing
```

Comment:
This worries me a little because we can add a new built-in type and forget to add it here, and we won't get any kind of warning from the compiler.  I don't immediately see any way of avoiding that though.  Maybe we could have a warning comment somewhere that you'd see it when you're adding a new type, for example at the definition of `mkNil` or `DefaultUni` ?  Or could we have a test that iterates over everything in `DefaultUni` and checks that `mkNil` works for the Plinth version?

Reply:
> This worries me a little because we can add a new built-in type and forget to add it here

@SeungheonOh add

```haskell
{-| Test that 'mkNil' works for all types from 'PLC.DefaultUni'
-}
instance
  TestAllBuiltinsSatisfy
    (PLC.AllBuiltinArgs PLC.DefaultUni MkNil)
    MkNil
```

to `PlutusTx.Lift.TestInstances` and you'll find out that there are still built-in types missing. Also add a comment saying that if an instance for a built-in type is missing, one should go and update the `tyUniverse` thing.